File: script/kubernetes/smoke/reporter.rb
# frozen_string_literal: true

module KubernetesSmoke
  class CheckFailure < StandardError; end

  class CheckSkip < StandardError; end

  class Reporter
    attr_reader :failures, :passes, :skips, :warnings

    def initialize(io, verbose: false)
      @io = io
      @verbose = verbose
      @failures = []
      @passes = []
      @skips = []
      @warnings = []
      @section = nil
    end

    def section(title)
      return if @section == title

      @section = title
      @io.puts
      @io.puts "#{title}:"
    end

    def check(name, required: true)
      detail = yield
      pass(name, detail)
    rescue CheckSkip => error
      skip(name, error.message)
    rescue CheckFailure => error
      if required
        fail(name, error.message)
      else
        warn(name, error.message)
      end
    rescue Kubectl::CommandError => error
      message = command_error_message(error)
      if required
        fail(name, message)
      else
        warn(name, message)
      end
    rescue StandardError => error
      message = "#{error.class}: #{error.message}"
      if required
        fail(name, message)
      else
        warn(name, message)
      end
    end

    def pass(name, detail = nil)
      @passes << [name, detail]
      line("PASS", name, detail)
    end

    def fail(name, detail)
      @failures << [name, detail]
      line("FAIL", name, detail)
    end

    def warn(name, detail)
      @warnings << [name, detail]
      line("WARN", name, detail)
    end

    def skip(name, detail)
      @skips << [name, detail]
      line("SKIP", name, detail)
    end

    def info(message)
      @io.puts "  INFO #{message}" if @verbose
    end

    def command(command)
      info("$ #{command.join(' ')}")
    end

    def summary
      @io.puts
      @io.puts "Summary:"
      @io.puts "  passes: #{passes.length}"
      @io.puts "  warnings: #{warnings.length}"
      @io.puts "  skips: #{skips.length}"
      @io.puts "  failures: #{failures.length}"
      return if failures.empty?

      @io.puts
      @io.puts "Failures:"
      failures.each do |name, detail|
        @io.puts "  - #{name}: #{detail}"
      end
    end

    def exit_code
      summary
      failures.empty? ? 0 : 1
    end

    private

    def line(status, name, detail)
      if detail.nil? || detail.to_s.empty?
        @io.puts "  #{status} #{name}"
      else
        @io.puts "  #{status} #{name} - #{detail}"
      end
    end

    def command_error_message(error)
      parts = ["#{error.command.join(' ')} exited #{error.status.exitstatus}"]
      stderr = error.stderr.to_s.strip
      stdout = error.stdout.to_s.strip
      parts << "stderr: #{stderr}" unless stderr.empty?
      parts << "stdout: #{stdout}" unless stdout.empty?
      parts.join("; ")
    end
  end
end

File: script/kubernetes/smoke-test/reporter.go
package main

import (
	"errors"
	"fmt"
	"io"
	"strings"
)

type CheckFailure string

func (err CheckFailure) Error() string {
	return string(err)
}

type CheckSkip string

func (err CheckSkip) Error() string {
	return string(err)
}

type CheckResult struct {
	Name   string
	Detail string
}

type Reporter struct {
	io       io.Writer
	verbose  bool
	section  string
	Passes   []CheckResult
	Failures []CheckResult
	Warnings []CheckResult
	Skips    []CheckResult
}

func NewReporter(io io.Writer, verbose bool) *Reporter {
	return &Reporter{io: io, verbose: verbose}
}

func (reporter *Reporter) Section(title string) {
	if reporter.section == title {
		return
	}
	reporter.section = title
	fmt.Fprintln(reporter.io)
	fmt.Fprintf(reporter.io, "%s:\n", title)
}

func (reporter *Reporter) Check(name string, required bool, fn func() (string, error)) {
	detail, err := fn()
	if err == nil {
		reporter.Pass(name, detail)
		return
	}

	var checkSkip CheckSkip
	if errors.As(err, &checkSkip) {
		reporter.Skip(name, checkSkip.Error())
		return
	}

	var commandError *CommandError
	if errors.As(err, &commandError) {
		reporter.recordError(name, commandError.Message(), required)
		return
	}

	var checkFailure CheckFailure
	if errors.As(err, &checkFailure) {
		reporter.recordError(name, checkFailure.Error(), required)
		return
	}

	reporter.recordError(name, fmt.Sprintf("%T: %v", err, err), required)
}

func (reporter *Reporter) Pass(name string, detail string) {
	reporter.Passes = append(reporter.Passes, CheckResult{Name: name, Detail: detail})
	reporter.line("PASS", name, detail)
}

func (reporter *Reporter) Fail(name string, detail string) {
	reporter.Failures = append(reporter.Failures, CheckResult{Name: name, Detail: detail})
	reporter.line("FAIL", name, detail)
}

func (reporter *Reporter) Warn(name string, detail string) {
	reporter.Warnings = append(reporter.Warnings, CheckResult{Name: name, Detail: detail})
	reporter.line("WARN", name, detail)
}

func (reporter *Reporter) Skip(name string, detail string) {
	reporter.Skips = append(reporter.Skips, CheckResult{Name: name, Detail: detail})
	reporter.line("SKIP", name, detail)
}

func (reporter *Reporter) Info(message string) {
	if reporter.verbose {
		fmt.Fprintf(reporter.io, "  INFO %s\n", message)
	}
}

func (reporter *Reporter) Command(command []string, dryRun bool) {
	if dryRun {
		fmt.Fprintf(reporter.io, "  DRY-RUN %s\n", strings.Join(command, " "))
		return
	}
	reporter.Info("$ " + strings.Join(command, " "))
}

func (reporter *Reporter) Summary() {
	fmt.Fprintln(reporter.io)
	fmt.Fprintln(reporter.io, "Summary:")
	fmt.Fprintf(reporter.io, "  passes: %d\n", len(reporter.Passes))
	fmt.Fprintf(reporter.io, "  warnings: %d\n", len(reporter.Warnings))
	fmt.Fprintf(reporter.io, "  skips: %d\n", len(reporter.Skips))
	fmt.Fprintf(reporter.io, "  failures: %d\n", len(reporter.Failures))
	if len(reporter.Failures) == 0 {
		return
	}

	fmt.Fprintln(reporter.io)
	fmt.Fprintln(reporter.io, "Failures:")
	for _, failure := range reporter.Failures {
		fmt.Fprintf(reporter.io, "  - %s: %s\n", failure.Name, failure.Detail)
	}
}

func (reporter *Reporter) ExitCode() int {
	reporter.Summary()
	if len(reporter.Failures) == 0 {
		return 0
	}
	return 1
}

func (reporter *Reporter) line(status string, name string, detail string) {
	if detail == "" {
		fmt.Fprintf(reporter.io, "  %s %s\n", status, name)
		return
	}
	fmt.Fprintf(reporter.io, "  %s %s - %s\n", status, name, detail)
}

func (reporter *Reporter) recordError(name string, detail string, required bool) {
	if required {
		reporter.Fail(name, detail)
		return
	}
	reporter.Warn(name, detail)
}
